src/ext.rs

```diff
@@ -0,0 +1,444 @@
+        }
+        SupportedType::Bytes => serializer.serialize(obj.downcast::<PyBytes>()?.as_bytes()),
+        SupportedType::Str => serializer.serialize(&obj.downcast::<PyString>()?.to_cow()?),
+        SupportedType::Int8 => serializer.serialize(i8::extract_bound(obj)?),
+        SupportedType::Int16 => serializer.serialize(i16::extract_bound(obj)?),
+        SupportedType::Int32 => serializer.serialize(i32::extract_bound(obj)?),
+        SupportedType::Int64 => serializer.serialize(i64::extract_bound(obj)?),
+        SupportedType::Int128 => serializer.serialize(i128::extract_bound(obj)?),
+        SupportedType::UInt8 => serializer.serialize(u8::extract_bound(obj)?),
+        SupportedType::UInt16 => serializer.serialize(u16::extract_bound(obj)?),
+        SupportedType::UInt32 => serializer.serialize(u32::extract_bound(obj)?),
+        SupportedType::UInt64 => serializer.serialize(u64::extract_bound(obj)?),
+        SupportedType::UInt128 => serializer.serialize(u128::extract_bound(obj)?),
+        SupportedType::Float | SupportedType::Float64 => {
+            serializer.serialize(f64::extract_bound(obj)?)
+        }
+        SupportedType::Float32 => serializer.serialize(f64::extract_bound(obj)? as f32),
+        SupportedType::VarInt | SupportedType::Int => {
+            serializer.serialize(VarInt(i64::extract_bound(obj)?))
+        }
+        SupportedType::VarUInt => serializer.serialize(VarInt(u64::extract_bound(obj)?)),
+        SupportedType::Bool => serializer.serialize(bool::extract_bound(obj)?),
+        SupportedType::List => serialize_iter(
+            serializer,
+            obj.downcast::<PyList>()?,
+            item_type,
+            serialize_item,
+        )?,
+        SupportedType::Tuple => {
+            let tuple = obj.downcast::<PyTuple>()?;
+            for item in tuple {
+                serialize(serializer, &item)?;
+            }
+        }
+        SupportedType::Dict => serialize_iter(
+            serializer,
+            obj.downcast::<PyDict>()?,
+            pair_type,
+            serialize_pair,
+        )?,
+        SupportedType::Set => serialize_iter(
+            serializer,
+            obj.downcast::<PySet>()?,
+            item_type,
+            serialize_item,
+        )?,
+        SupportedType::FrozenSet => serialize_iter(
+            serializer,
+            obj.downcast::<PyFrozenSet>()?,
+            item_type,
+            serialize_item,
+        )?,
+    }
+    Ok(())
+}
+
+fn serialize_iter<T, I: IntoIterator<Item = T>, Ty: Eq + Copy>(
+    serializer: &mut ZSerializer,
+    iter: I,
+    get_type: impl Fn(&T) -> PyResult<Ty>,
+    serialize: impl Fn(&mut ZSerializer, T, Ty) -> PyResult<()>,
+) -> PyResult<()>
+where
+    I::IntoIter: ExactSizeIterator,
+{
+    let iter = iter.into_iter();
+    serializer.serialize(VarInt(iter.len()));
+    let mut tp = None;
+    for item in iter {
+        match &tp {
+            Some(tp) if get_type(&item)? != *tp => {
```

Comment:
will this check work for list of lists of different types ?

Reply:
No, it will not. I didn't push to that edge case, but I can if you think it's necessary.